text-editor/text-editor-002-en-proofing.md

```diff
@@ -0,0 +1,1015 @@
+    catchError(() => of(error(action.editorName))),
+  )),
+)
+
+// insertTweetNode :: (Object, State.TextEditor) -> Promise
+const insertTweetNode = ({ editorName, url, html }, textEditor) => new Promise(resolve => {
+  const tweetId = getTweetIdFromUrl(url);
+  const uid = uniqid(tweetId);
+
+  const newNode = pipe(
+    tap(e => e.innerHTML = renderToString(UnconnectedTweet(tweetId, uid), 'text/html')),
+    prop('firstChild'),
+  )(document.createElement('div'));
+
+  insertNewNodeAtIndex(
+    newNode,
+    textEditor.ParagraphToolbox[editorName].targetNodeIndex,
+    editorName,
+  );
+
+  const originalHtmlMarkup = pipe(
+    tap(e => e.innerHTML = html),
+    // only grab the `blockquote` tag, not the `script` tag
+    e => e.firstChild.outerHTML,
+  )(document.createElement('div'));
+
+  resolve([ editorName, tweetId, uid, originalHtmlMarkup ]);
+})
+
+// insertNewNodeAtIndex :: (Element, Number, String) -> _
+export const insertNewNodeAtIndex = (newNode, targetIndex, editorName) =>
+  getEditor(editorName).replaceChild(
+    newNode,
+    nth(targetIndex, getRootNodesAsArray(editorName)),
+  );
+
+// getRootNodesAsArray :: String -> [Node]
+export const getRootNodesAsArray = editorName => Array.from(
+  getEditor(editorName).childNodes,
+)
+
+// getEditor :: String -> Node
+export const getEditor = editorName => document.querySelector(
+  `.edited-text-root[data-editor-name="${editorName}"]`
+)
+```
+[See gist here](https://gist.github.com/jaljo/4c8f83acc48766b930d7d584fe9ed22b)
+
+As you can see, the `insertTweetNode` function returns a Promise, it provide
+us a nice way to handle failures at that step of the process. What does that
+function do ?
+
+- Using a combination of react `renderToString` utility, one of our parsers, and
+the HTML markup provided by the twitter API in the previous observable, it
+creates an HTML representation of the tweet **formatted as we need it to be.**
+- Then, using an `insertNewNodeAtIndex` utility, we replace the empty paragraph
+from where the tweet insertion form was shown by the new node created at the
+previous step (that paragraph index has early been saved into the paragraph
+toolbox state when the `SHOW_PARAGRAPH_TOOLBOX` action has been reduced).
+- The promise is finally resolved with data such as the editor name, because we
+need to carry these information out during the whole process.
+
+Once the tweet has successfully been inserted, a `TWEET_INSERTED` action is
+dispatched by the epic. Otherwise, an error is catched and displayed to the
+user. Notice that the kind of error we handle here is not of the same type than
+in the previous observable. Errors that may occur here could only be DOM
+related.
+
+### Rendering the inserted tweet and going back to the original state
+
+The `TWEET_INSERTED` action is observed by four epics (and here we gain full
+benefit of working with observable streams) :
+- One that takes care of systematically insert and focus a new paragraph after a
```

Comment:
One that takes care of systematically insert and focus a new paragraph after a -> One that systematically takes care of inserting and focusing a new paragraph after a